[ERE-1954] Longitudinal followups

ScriptSmith opened the pull request:
After the merge, it needs the accompanying ansible project to deploy the stack, and some more testing in the deployed environment.

Also need to update the registry repos & cookiecutter template.

ppettitau committed:
* check recipient has not unsubscribed from email notification before sending it

ppettitau committed:
[ERE-1949] Unsubscribe from Email Notifications
* Add email preferences management
* Add option to unsubscribe from all emails
* Add util to generate a token for a given user, use the token to create URLs and check the token for validity

ppettitau committed:
[ERE-1949] Customise send_mail to allow List-Unsubscribe header to be set

ScriptSmith commented:
Updating the test zips:

```diff
diff --git a/rdrf/rdrf/fixtures/longitudinal_followups.yaml b/rdrf/rdrf/fixtures/longitudinal_followups.yaml
new file mode 100644
index 000000000..e00256952
--- /dev/null
+++ b/rdrf/rdrf/fixtures/longitudinal_followups.yaml
@@ -0,0 +1,131 @@
+- model: rdrf.longitudinalfollowup
+  pk: 1
+  fields:
+    name: Followup 1
+    description: Followup 1
+    context_form_group: 2
+    frequency: 1 00:00:00
+    debounce: 7 00:00:00
+    condition: ''
+- model: rdrf.longitudinalfollowup
+  pk: 2
+  fields:
+    name: Followup 2
+    description: Followup 2
+    context_form_group: 3
+    frequency: 1 00:00:00
+    debounce: 7 00:00:00
+    condition: ''
+- model: rdrf.emailtemplate
+  pk: 1
+  fields:
+    language: en
+    description: Followup
+    subject: Followup
+    body: "<p>\r\nHello {{patient.given_names}} {{patient.family_name}},\r\n</p>\r\n\r\n<p>\r\nYou
+      are due to follow up on the following forms in the registry:\r\n<ul>\r\n{% for
+      name, entries in longitudinal_followups.items %}\r\n<li><a href=\"{{ entries.0.longitudinal_followup.context_form_group.link
+      \ }}\">{{ name }}</a></li>\r\n{% endfor %}\r\n</ul>\r\n<p>\r\n\r\n<p>\r\nRegards,\r\n</p>\r\n<p>\r\nCurator\r\n</p>"
+- model: rdrf.emailnotification
+  pk: 1
+  fields:
+    description: longitudinal-followup
+    public_description: Longitudinal followup
+    registry: 2
+    email_from: no-reply@registryframework.net
+    recipient: curator@example.com
+    group_recipient: null
+    disabled: false
+    subscribable: true
+    email_templates:
+    - 1
+    file_uploaded_cdes: []
+- model: rdrf.contextformgroup
+  pk: 3
+  fields:
+    registry: 2
+    context_type: M
+    code: CFG3
+    name: Checkup2
+    abbreviated_name: CFG3
+    naming_scheme: D
+    is_default: false
+    naming_cde_to_use: SecondFollowUp/FHFollowUp/DateOfAssessment
+    ordering: C
+    sort_order: 2
+- model: rdrf.registryform
+  pk: 7
+  fields:
+    registry: 2
+    name: SecondFollowUp
+    abbreviated_name: SecondFollowUp
+    display_name: Second Follow Up
+    header: ''
+    sections: FHFollowUp,FHEvent,FHFollowUpHypertension,FHFollowUpDiabetes,FHFollowUpAntithrombotic,FHBiochemistry,FHDeath
+    position: 6
+    applicability_condition: ''
+    conditional_rendering_rules: ''
+    tags: '[]'
+    save_position: false
+    complete_form_cdes: [ ]
+    groups_allowed: [ ]
+    groups_readonly: [ ]
+- model: rdrf.contextformgroupitem
+  pk: 7
+  fields:
+    context_form_group: 3
+    registry_form: 7
+- model: rdrf.registry
+  pk: 2
+  fields:
+    name: FH Registry
+    code: fh
+    desc: fh
+    splash_screen: "<h1 style=\"color: #000066\"><p class=\"text-center\">National
+      Registry for \r\nFamilial Hypercholesterolemia</p><h1 style=\"color: #000066\">\r\n\r\n<h1
+      style=\"color: #000066\">Background</h1>\r\n<p class=\"text-justify\">\r\nFamilial
+      hypercholesterolemia (FH) is an autosomal dominant disorder of absent or grossly
+      malfunctioning low-density lipoprotein (LDL) receptors that causes severe elevations
+      in total cholesterol and low-density lipoprotein cholesterol (LDLc). Heterozygous
+      FH occurs in approximately 1 per 500 persons.\r\n</p>\r\n\r\n<p class=\"text-justify\">\r\nFor
+      the purposes of this Registry, Familial Hypercholesterolemia will be inclusive
+      of other forms of autosomal dominant hypercholesterolemia (FDB, PCSK9 gain-of-function
+      mutations). \r\n</p>\r\n\r\n<p class=\"text-justify\">\r\nFH is associated with
+      a high risk for premature coronary artery disease. Early detection and aggressive
+      management to lower the LDLc level helps prevent or slows the progression of
+      coronary atherosclerosis. Moreover, if the first-degree relatives of a patient
+      with FH are screened, other gene carriers can be identified and treated.\r\n</p>\r\n\r\n<h1
+      style=\"color: #000066\">Purpose</h1>\r\n<p class=\"text-justify\">\r\nThe purpose
+      of the National Registry for Familial Hypercholesterolemia is to collate data
+      to facilitate clinical service planning and to inform clinical best practice.
+      The Registry will also enable research on aggregated data and the identification
+      of eligible volunteers for clinical trials.\r\n</p>\r\n\r\n<h1 style=\"color:
+      #000066\">Objectives</h1>\r\n<p class=\"text-justify\">\r\nTo facilitate service
+      planning by analyses and reporting of data collected by the Registry on prevalence,
+      geographical distribution, genetic variants associated with disease, clinical
+      features, clinical management and patient outcomes.\r\n</p>\r\n\r\n<p class=\"text-justify\">\r\nTo
+      enable research by providing aggregate, de-identified data to research entities.\r\n</p>\r\n\r\n<p
+      class=\"text-justify\">\r\nTo facilitate identification and recruitment of eligible
+      volunteers for clinical trials.\r\n</p>\r\n\r\n<p class=\"text-justify\">\r\nTo
+      promulgate new knowledge to inform best practice and care services. \r\n</p>\r\n\r\n<h1
+      style=\"color: #000066\">Administration/Oversight</h1>\r\n<p class=\"text-justify\">\r\nThe
+      National Registry for Familial Hypercholesterolemia is governed by a National
+      Advisory Board. The Board includes representatives from each jurisdiction from
+      within the Familial Hypercholesterolemia Australasia Network and will oversee
+      all registry activities including all enquiries for access. \r\n</p>\r\n\r\n<p
+      class=\"text-justify\">\r\n<a href=\"https://fhregistry-international.com/pdf/FHAN%20AAS%20Charter%20National%20FH%20Registry%202015_Endorsed%2010.09.2015.pdf\">The
+      Charter for the National Familial Hypercholesterolaemia Registry</a> and the
+      <a href=\"https://fhregistry-international.com/pdf/FH%20Registry%20Brochure_v5.pdf\">Registry
+      Brochure</a> are available for your information. \r\n\r\n\r\n</p>\r\n\r\n<h1
+      style=\"color: #000066\">Contact</h1>\r\n<p class=\"text-justify\">\r\nFor more
+      information on the National Familial Hypercholesterolaemia Registry, please
+      contact <a href=\"mailto:fhwa@health.wa.gov.au\">fhwa@health.wa.gov.au.</a>
+      \r\n\r\n</p>\r\n\r\n<p class=\"text-justify\">\r\nA sample consent form for
+      patients wishing to participate in the registry can be downloaded  <a href=\"https://fhregistry-international.com/pdf/ADULT%20PICF%20Version%201.0.pdf\">here.</a>\r\n\r\n</p>\r\n\r\n<br/>\r\n<br/>\r\n<br/>\r\n<br/>"
+    version: 0.1.53
+    patient_data_section: 40
+    metadata_json: '{"visibility": ["PatientRelativeForm"], "features": ["family_linkage",
+      "contexts", "longitudinal_followups"], "family_linkage_form_name": "ClinicalData",
+      "family_linkage_section_code": "fhDateSection", "family_linkage_cde_code": "CDEIndexOrRelative",
+      "family_linkage_index_value": "fh_is_index", "family_linkage_relative_value":
+      "fh_is_relative"}'
diff --git a/scripts/update-test-zips.sh b/scripts/update-test-zips.sh
index f15f675dc..b014117d8 100755
--- a/scripts/update-test-zips.sh
+++ b/scripts/update-test-zips.sh
@@ -28,6 +28,7 @@ export_zip() {
     set +e
 
     teststack-exec-django import "${1}" --force
+    read -n1 -s -r -p $'Press space to continue...\n' key
     teststack-exec-django export registry --filename "${1}" --registry-code "${2}"
 
     set -e
@@ -35,6 +36,6 @@ export_zip() {
     teststack stop
 }
 
-export_zip /app/rdrf/rdrf/testing/behaviour/features/exported_data/dd.zip dd
+#export_zip /app/rdrf/rdrf/testing/behaviour/features/exported_data/dd.zip dd
 export_zip /app/rdrf/rdrf/testing/behaviour/features/exported_data/fh.zip fh
-export_zip /app/rdrf/rdrf/testing/behaviour/features/exported_data/ang.zip ang
+#export_zip /app/rdrf/rdrf/testing/behaviour/features/exported_data/ang.zip ang
```

```
./scripts/update-test-zips.sh
# wait for pause
docker-compose -f docker-compose-teststack-base.yml -f docker-compose-teststack-dev.yml exec serverundertest /docker-entrypoint.sh django-admin loaddata longitudinal_followups.yaml
# continue
```

ScriptSmith commented:
Example email template:

```django
<p>
Hello {{patient.given_names}} {{patient.family_name}},
</p>

<p>
You are due to follow up on the following forms in the registry:
<ul>
{% for name, entries in longitudinal_followups.items %}
<li><a href="{{ entries.0.longitudinal_followup.context_form_group.link }}">{{ name }}</a></li>
{% endfor %}
</ul>
<p>

<p>
Regards,
</p>
<p>
Curator
</p>
```

ppettitau commented on lambda.py:
```diff
@@ -0,0 +1,43 @@
+def _set_env_variables(parameter_envs):
+    ssm = boto3.client("ssm")
+    response = ssm.get_parameters(Names=list(parameter_envs.keys()), WithDecryption=True)
+    logger.debug(f"SSM parameters: {response}")
```
Will this debug statement reveal our secrets / passwords in plain text?

ScriptSmith replied:
Should be OK as the default logging level is `WARNING`: https://docs.python.org/3/howto/logging.html#when-to-use-logging (ie. debug statements aren't logged unless we change the logging level)

ppettitau commented on rdrf/rdrf/forms/email_preferences.py:
```diff
@@ -0,0 +1,83 @@
+                if self.instance.unsubscribe_all or self.instance.is_email_allowed(email_notification)]
+        else:
+            # For users without any existing saved preferences,
+            # select option to choose email subscriptions, with each, notification selected by default.
```
nitpick, fix grammar (probably my fault anyway): `select option to choose email subscriptions, with each, notification selected by default.`

ScriptSmith replied:
Changed

ppettitau commented on rdrf/rdrf/helpers/utils.py:
```diff
@@ -864,7 +864,7 @@ def validate_abbreviated_name(value):
     splitted = urlsplit(relative_url)
     domain = Site.objects.get_current().domain.rstrip('/')
-    scheme = 'https' if domain != 'localhost:8000' else 'http'
+    scheme = 'https' if domain != 'localhost:8000' and domain != 'serverundertest:8000' else 'http'
```
nitpick: Could have used `not in` instead of extra condition

ScriptSmith replied:
Updated

ppettitau commented on rdrf/rdrf/services/io/notifications/longitudinal_followups.py:
```diff
@@ -0,0 +1,163 @@
+        for entry in patient_entries:
+            entry.sent_at.append(sent_at)
+            entry.state = LongitudinalFollowupQueueState.SENT
+        LongitudinalFollowupEntry.objects.bulk_update(patient_entries, ["sent_at", "state"])
```
Should this be done after process_notification in case the `process_notification` fails?

ScriptSmith replied:
I believe we discussed this in a dev meeting, considering the two failure modes:

1. An email is successfully sent to a user's address, but there's an error on our end recording that success
2. An email fails to send to a user's address, and we record it as being sent when it hasn't

The suggestion was to favor the latter, because the former could lead to repeatedly sending emails unwanted emails to a user.

Initially, I had a third `LongitudinalFollowupQueueState` called `FAILED` which would allow admins to handle send failutes, but I believe it was suggested to just use two queue states and accept that some emails don't get delivered, in order to reduce complexity.

But worth revisiting this again now that we're about to merge.

ppettitau replied:
It looks like `process_notification` returns `True/False` depending if the notification was actually processed. One of the scenarios where it would return False is where the notification has had its emails "disabled". In this instance, I don't think we'd want to mark the follow up as successful / sent, would we? Since there has been no attempt to even send the email.

So I guess it's fair enough not to assume an exception means that an email hasn't been sent, but if the process_notification returns false then we should be pretty certain that an email wasn't even attempted.

ScriptSmith replied:
Closing as per discussion in dev meeting. In the documentation, I'll include a note about this.

ppettitau commented on rdrf/rdrf/templates/email_preference/_resubscribe_instructions.html:
```diff
@@ -0,0 +1,4 @@
+{% load i18n %}
+{% load project_title %}
+
+<p>{% trans "You may choose to resubscribe in the future by selecting the 'Email Preferences' option form the User menu in the " %}{% project_title %}.</p>
```
Spelling error: "form" should be "from"

ScriptSmith replied:
Fixed

ppettitau commented on rdrf/rdrf/testing/behaviour/features/longitudinal_followups/fh_followup_email.feature:
```diff
@@ -0,0 +1,81 @@
+    And I press "Add" button in "Follow Up" group in sidebar
+    And I enter value "01-01-2023" for form "Follow Up" section " " cde "Date of assessment"
+    And I click the "Save" button
+      And I press "Add" button in "Second Follow Up" group in sidebar
```
Indentation

ScriptSmith replied:
Fixed

ppettitau commented on rdrf/rdrf/testing/behaviour/features/longitudinal_followups/fh_followup_email.feature:
```diff
@@ -0,0 +1,81 @@
+    And send emails 48 hours later
+    Then I see 1 email
+
+    When I unsubscribe in email 1
```
Awkwardly worded.

ppettitau commented on rdrf/rdrf/models/definition/models.py:
```diff
@@ -1172,6 +1180,7 @@ class EmailNotification(models.Model):
         (EventType.CLINICIAN_ASSIGNED, "Clinician Assigned"),
         (EventType.CLINICIAN_UNASSIGNED, "Clinician Unassigned"),
         (EventType.FILE_UPLOADED, "File Uploaded"),
+        (EventType.LONGITUDINAL_FOLLOWUP, "Longitudinal Followup"),
```
For consistency, add feature requirements in square brackets similar to other email notifications that require a feature to be enabled.

![image](https://user-images.githubusercontent.com/60163872/235812675-f0ab8d49-70e1-4b62-b283-7533d2cd8e0e.png)

ScriptSmith replied:
Done

ppettitau commented on rdrf/registry/patients/admin.py:
```diff
@@ -545,6 +547,23 @@ def membership(self, patient_model):
+
+class LongitudinalFollowupEntryAdmin(admin.ModelAdmin):
+    model = LongitudinalFollowupEntry
+    list_display = (
```
Do all these fields need to be in the list? The columns seem a bit cramped:
![image](https://user-images.githubusercontent.com/60163872/235814507-ace6ad58-a9bb-459a-b77a-7e16adde0cda.png)

ScriptSmith replied:
I think so yeah. Which would you remove?

ppettitau commented on rdrf/rdrf/testing/behaviour/features/longitudinal_followups/fh_followup_email.feature:
```diff
@@ -0,0 +1,81 @@
+    Given a registry named "FH Registry" with code "fh"
+    Given I am expecting to receive email
+
+  Scenario: Receive a followup email
```
These scenarios are very 'script'-like and don't really describe the behaviour of the feature.

Would be nice to have something a bit more natural language e.g
```
  Scenario: Receive a followup email
  Given "SMITH John" added an entry for form "Follow Up" today
  And a longitudinal follow up is scheduled to run every 24 hours
  When 1 hour has passed
  Then no emails are sent
  When 24 hours has passed
  Then 1 email is sent
  And email 1 has 1 link
  And each link in email 1 loads successfully
  When I empty the mail
  And 48 hours has passed
  Then no emails are sent
```

ScriptSmith replied:
I initially wrote parts of the tests in a more succinct style, but was finding that I just re-implemented a lot of the functionality of the existing BDD tests, and as per discussions in the dev meeting about not refactoring those 'legacy' tests, I elected to just add a few small steps to make my tests work.

But I'll give it another shot :+1:

ppettitau commented on rdrf/rdrf/testing/behaviour/features/longitudinal_followups/steps.py:
```diff
@@ -0,0 +1,54 @@
+@step(r'see (\d+) emails?')
+def check_mail(_step, count):
+    go_to_url("mail/outbox")
+    sent_emails = len(world.browser.find_elements_by_css_selector("#id_messages > tbody > tr"))
```
Selectors could be refactored to reduce duplication.

ScriptSmith replied:
Done

ppettitau commented on rdrf/rdrf/testing/behaviour/features/longitudinal_followups/page.py:
Empty file. If it's not needed, delete it.

ScriptSmith replied:
Done

ppettitau commented on rdrf/rdrf/testing/unit/longitudinal_followup_tests.py:
Would be good to have a test case that tests `condition` on the Longitudinal Followup

ScriptSmith replied:
Added

ppettitau commented on rdrf/rdrf/views/email_preferences_view.py:
Is this working for you? I get an error when I attempt use the unsubscribe link.

ScriptSmith replied:
Double checked and both links work for me

ppettitau commented:
It would be helpful to have a 'Getting Started' or user guide on what needs to be configured for this to work. I've been thinking about how we could possibly write doco for what we deliver and somehow form part of the wider documentation for TRRF but I don't have any good ideas on how to go about this without it being very piecemeal.

ppettitau commented on rdrf/rdrf/admin.py:
```diff
@@ -452,6 +453,23 @@ class DashboardWidgetAdmin(admin.ModelAdmin):
     inlines = [DashboardLinksInline, DashboardDemographicsInline, DashboardCdeDataInline]
 
 
+class LongitudinalFollowupAdmin(admin.ModelAdmin):
```
Can we use the feature flags to hide the admin options for Longitudinal Followup and Longitudinal Followup Entries?

ScriptSmith replied:
Done

ppettitau commented:
How will we create longitudinal follow up entries for existing patients when a new longitudinal followup is created?

ppettitau commented:
On behalf of Ryan: We should have a kill switch for bulk send.

My notes:
- Looks like there already is one: RegistryFeatures.LONGITUDINAL_FOLLOWUPS
- But should we check it each time we iterate through each patient entry in the loop responsible for sending emails?

ScriptSmith commented:
> How will we create longitudinal follow up entries for existing patients when a new longitudinal followup is created?

https://github.com/eresearchqut/trrf/blob/4dcce8685600e5ecacb338caafaa12a443b75bb0/rdrf/rdrf/management/commands/create_longitudinal_followup_entries.py

> But should we check it each time we iterate through each patient entry in the loop responsible for sending emails?

Yeah I suppose. Shouldn't cause any undue load on the database.

ppettitau commented on rdrf/rdrf/admin.py:
```diff
@@ -452,6 +453,23 @@ class DashboardWidgetAdmin(admin.ModelAdmin):
+class LongitudinalFollowupAdmin(admin.ModelAdmin):
+    model = LongitudinalFollowup
+
+    list_display = (
```
Is it worth adding some validation to ensure that an email notification exists for the registry that relates to the context form group selected when creating/editing a `LongitudinalFollowup`?

ScriptSmith replied:
Closing as per discussion in dev meeting.

ScriptSmith commented on rdrf/rdrf/management/commands/create_longitudinal_followup_entries.py:
```diff
@@ -0,0 +1,71 @@
+
+            return True
+
+        entries = LongitudinalFollowupEntry.objects.bulk_create(
```
Should prompt for confirmation before creation

ScriptSmith replied:
Done

ScriptSmith committed:
Delete empty file

ScriptSmith committed:
Spelling + grammar + syntax fixes

ScriptSmith committed:
Conditionally show LongitudinalFollowups on admin page

ScriptSmith committed:
Simplify local domain check

ScriptSmith committed:
Prompt before creating LFEs

ScriptSmith committed:
Wrapping ConditionException

ScriptSmith committed:
Check for longitudinal_followups feature before sending each email

ScriptSmith committed:
Simplify email selector

ppettitau commented on rdrf/rdrf/services/io/notifications/longitudinal_followups.py:
```diff
@@ -135,6 +148,11 @@ def send_longitudinal_followups(now):
 
+        patient_registry = patient.rdrf_registry.first()
+
+        if not patient_registry.has_feature(RegistryFeatures.LONGITUDINAL_FOLLOWUPS):
```
Is there a better way to do this? Is it worth having a button somewhere that just causes this process to exit completely? (rather than skip to the next iteration).

I don't have a good answer to this.

ScriptSmith replied:
There's no way to stop a running lambda, so I think something similar to this is the only way.

ppettitau replied:
I mean using `break` instead of `continue`, to exit the loop entirely.

ScriptSmith replied:
Yeah I think that works

ScriptSmith committed:
Halt lf execution when feature toggles off

ScriptSmith committed:
Simplify lf e2e tests

ppettitau commented on rdrf/rdrf/testing/behaviour/features/longitudinal_followups/fh_followup_email.feature:
```diff
@@ -7,75 +7,58 @@ Feature: Longitudinal followup email notifications
+    When 24 hours pass
+    Then 1 email is sent
+
+    Then the mail is emptied
```
`Then` should only be used for assertions, so you could either cleanup the mailbox after you make the previous assertion of how many emails have been sent.. or you could change this to a `Given`, or possibly a `When`

ScriptSmith replied:
As I can see it, we can:

1. Change the keyword from `Then`
    a. Happy to do this, but the grammar feels a bit awkward. Can I just use `And`?
3. Add it to the `links_load` step, then make the 3rd scenario longer to test for this
    a. Adding extra steps to the scenario is the opposite of what we're going for.
5. Don't empty the emails at all, and keep a global count of 'read' emails in the outbox
    a. Keeping track of state throughout the tests makes debugging harder.

ppettitau replied:
1. If you use `And` after a `Then`, it's still part of the assertion block. But you can move it down to the `When` and make use of `And` there. 
e.g.:
```
    When 24 hours pass
    Then 1 email is sent

    When the mail is emptied
    And the "Follow Up" form in patient "SMITH John" is filled out again
    And 48 hours pass
    Then 1 email is sent
    ...
```
2. You could, but then you'd be checking `links_load` each time when you're not currently already doing that. 
3. Agree this option makes everything harder.

ScriptSmith replied:
Ok, I'll go with option 1

ScriptSmith committed:
Use better Gherkin syntax
'When' for actions
'Then' for assertions
'And' to extend a block of actions or assertions

ppettitau committed:
Fix failing test

ScriptSmith committed:
Add longitudinal followups

ScriptSmith committed:
Update longitudinal followup email label
